src/analysis/special_functions/bernstein.lean

```diff
@@ -0,0 +1,408 @@
+open bernstein_approximation
+open bounded_continuous_function
+
+/--
+This is the proof given in [Richard Beals' *Analysis, an introduction*][beals-analysis], §7D,
+and reproduced on wikipedia.
+-/
+theorem bernstein_approximation_uniform (f : I →ᵇ ℝ) (ε : ℝ) (h : 0 < ε) :
+  ∃ n : ℕ, ∥bernstein_approximation n f - f∥ < ε :=
+begin
+  let δ := δ f (ε/2) (half_pos h),
+  let n : ℕ := _, use n, -- We postpone choosing `n` until we've obtained an explicit estimate.
+  suffices npos : 0 < (n : ℝ), -- However we do assume right away that it won't be `n = 0`!
+  -- Four easy inequalities we'll need later:
+  have w₀ : 0 ≤ ε / 2 := div_nonneg (le_of_lt h) (by norm_num),
+  have w₁ : 0 ≤ 2 * ∥f∥ := mul_nonneg (by norm_num) (norm_nonneg f),
+  have w₂ : 0 ≤ 2 * ∥f∥ * δ^(-2 : ℤ) := mul_nonneg w₁ (pow_minus_two_nonneg _),
+  have w₃ : ∀ x y, abs (f x - f y) ≤ 2 * ∥f∥ := λ x y,
+    calc abs (f x - f y) = abs (f x + -(f y)) : by rw sub_eq_add_neg
+      ... ≤ abs (f x) + abs (-f y) : abs_add _ _
+      ... = ∥f x∥ + ∥f y∥ : by rw [abs_neg, real.norm_eq_abs, real.norm_eq_abs]
+      ... ≤ ∥f∥ + ∥f∥ : add_le_add (norm_coe_le_norm _ _) (norm_coe_le_norm _ _)
+      ... = 2 * ∥f∥ : by ring,
+  -- As `[0,1]` is compact, it suffices to check the inequality pointwise.
+  apply bounded_continuous_function.norm_lt_of_compact,
+  intro x,
+  -- The idea is to split up the sum over `k` into two sets,
+  -- `S`, where `x - k/n < δ`, and its complement.
+  let S := S f ε h n x,
+  calc
+    abs ((bernstein_approximation n f - f) x)
+        = abs (bernstein_approximation n f x - f x)
+                              : rfl
+    ... = abs (bernstein_approximation n f x - f x * 1)
+                              : by rw mul_one
+    ... = abs (bernstein_approximation n f x - f x * (∑ k : fin (n+1), bernstein n k x))
+                              : by rw bernstein.probability
+    ... = abs (bernstein_approximation n f x - (∑ k : fin (n+1), f x * bernstein n k x))
+                              : by rw finset.mul_sum
+    ... = abs (∑ k : fin (n+1), (f k/ₙ - f x) * bernstein n k x)
+                              : begin
+                                  dsimp [bernstein_approximation],
+                                  simp only [coe_sum, coe_smul, finset.sum_apply,
+                                    ←finset.sum_sub_distrib,
+                                    algebra.id.smul_eq_mul, ←sub_mul],
+                                end
+    ... ≤ ∑ k : fin (n+1), abs ((f k/ₙ - f x) * bernstein n k x)
+                              : finset.abs_sum_le_sum_abs
+    ... = ∑ k : fin (n+1), abs (f k/ₙ - f x) * bernstein n k x
+                              : by simp_rw [abs_mul, abs_eq_self.mpr bernstein_nonneg]
+    ... = ∑ k in S, abs (f k/ₙ - f x) * bernstein n k x +
+          ∑ k in Sᶜ, abs (f k/ₙ - f x) * bernstein n k x
+                              : (S.sum_add_sum_compl _).symm
+    -- We now work on the terms in `S`, where uniform continuity and `bernstein.probability`
```

Comment:
It might be a little easier to finish this calc block here, like

```lean
    ... < ε/2 + ε/2 : add_lt_add_of_le_of_lt _ _
    ... = ε : add_halves ε,
    { /- terms in S -/ },
    { /- terms in Sᶜ -/ }
```

That should save an `add_le_add_*` application per step, at least.

Reply:
Thank you! That's easier to read and write, and saves a little on inequality wrangling at the end.